server/src/lib/model/split.ts

```diff
@@ -0,0 +1,85 @@
+/**
+ * We're splitting the dataset into three buckets; train, dev and test. That's a common technique
+ * in ML: https://stats.stackexchange.com/a/19051
+ */
+
+export interface Split {
+  train: number;
+  dev: number;
+  test: number;
+}
+
+export const IDEAL_SPLIT: Split = Object.freeze({
+  train: 0.6,
+  dev: 0.2,
+  test: 0.2,
+});
+
+const sumValues = (split: Split): number =>
+  Object.values(split).reduce((sum, n) => sum + n, 0);
+
+export function normalize(split: Split) {
+  const totalCount = sumValues(split);
+  return Object.entries(split)
+    .map(([key, count]) => [key, count / totalCount])
+    .reduce((obj: any, [key, value]) => {
+      obj[key] = value;
+      return obj;
+    }, {});
+}
+
+/**
+ * Given a real split, in absolute numbers (i.e. count of occurrences) calculate a compensatory
+ * split which would move the real data closer to the split defined as IDEAL_SPLIT.
+ */
+export function calculateCompensatorySplit(realSplit: Split): Split {
+  const compensatorySplit: any = {};
+
+  for (const [bucket, count] of Object.entries(normalize(realSplit))) {
```

Comment:
We need some comments for each step of this function, since it's not obvious what is going on at first glance.